### ISSUE TITLE
MathJax breaks on \emph inside mathescape, round 2

### ISSUE THREAD

**Issue by henrikt-ma** [brucemiller/LaTeXML#1477]
Reduced example _mini.tex_:
```tex
\documentclass{report}

\usepackage{listings}

\lstset{
  mathescape=true,
}

\begin{document}

Works: \lstinline!foo($\langle\mbox{argument list}\rangle$)!

Using \lstinline!\emph! breaks Mathjax processing: \lstinline!foo($\langle\mbox{\emph{argument list}}\rangle$)!

Removing angle brackets works: \lstinline!foo($\mbox{\emph{argument list}}$)!

A display style listing:
\begin{lstlisting}
Same as above:
  foo($\langle\mbox{argument list}\rangle$)
  foo($\langle\mbox{\emph{argument list}}\rangle$)
  foo($\mbox{\emph{argument list}}$)
\end{lstlisting}

\end{document}
```

Build with:
```
mini.html: mini.tex
  latexml mini.tex --dest mini.xml
  latexmlpost mini.xml -format html --javascript=LaTeXML-maybeMathjax.js --dest $@
```

Mathjax fails on the `\emph` stuff, showing _Math input error_ with the tooltip _Unknown node type "em"_.  The problem appears to be the `<em>` tag:
```xml
<math id="p2.m1" class="ltx_Math" alttext="\langle\mbox{\emph{argument list}}\rangle" display="inline">
<mrow><mo stretchy="false">⟨</mo><mtext mathvariant="italic">
  <em class="ltx_emph ltx_font_italic">argument list</em>
 </mtext><mo stretchy="false">⟩</mo></mrow></math>
```

If the `<em>` tag wrapping the _argument list_ is removed, the error goes away.  I don't know MathML, but a quick search doesn't give me any indication that `<em>` would be a valid tag.

To see why removing the angle brackets work (mentioned here because of how we encountered this issue), note that the use of MathML goes away:
```xml
<span class="ltx_text ltx_markedasmath"><em class="ltx_emph ltx_font_italic">argument list</em></span>
```

**Comment by dginev** [comment]
Hi @henrikt-ma , this is a sort of a cross-over issue with the MathJax folks. The MathML specification, I am told, has always intended to allow HTML textual markup in the MathML use for HTML5 - and hence LaTeXML takes advantage of that specified feature.

In fact we have had this exact issue before at https://github.com/brucemiller/LaTeXML/issues/1323 . It is likely we'll have the same resolution here again, but maybe it's time someone rang the doorbell on the MathJax side? It must be easy for them to whitelist the element.

**Comment by henrikt-ma** [comment]
And you're sure the MathJax option is not important enough to consider LaTeXML workarounds for their shortcomings?

**Comment by xworld21** [comment]
For reference, this is the open [MathJax issue](https://github.com/mathjax/MathJax/issues/487) about using HTML inside `<mtext>`, dating back to 2013. They expect `<mtext>` to be pure text at the moment.

**Comment by dginev** [comment]
@henrikt-ma personally, I see this as a feature MathJax should support - they aim to be able to render valid MathML as well as valid TeX, and the example we are discussing is both of those. On top of that we have this example rendering well in Firefox already, and an expectation for the correct `mtext` behavior when MathML is enabled natively in the Chromium browsers.

In a way we are lucky MathJax supports as much of MathML as it does - if we were to instead consider limiting the MathML coverage to what is supported by say KaTeX, we wouldn't even be allowed to generate MathML - just to copy over the TeX strings for the formulas to the browser and hope KaTeX can handle them.

So while we don't have any desire to break compatibility with MathJax, I think some cases are better handled on their end rather than on ours.

**Comment by henrikt-ma** [comment]
I'm afraid you're right. :)

**Comment by brucemiller** [comment]
Indeed, pure, stand-alone, MathML should only have text in mtext;
but when embedded in some other document type, can contain the
(inline) markup of that type.

 From the logical point of view, your markup describes a mathematical
object which brackets an emphasized text object. Any workaround that
I can think of breaks that relationship.  But if you're willing to
give up that relationship, a source-level workaround would be:
```
$\langle$\emph{argument list}$\rangle$
```
Some day, maybe, a smart system might be able to reassemble it :>

**Comment by henrikt-ma** [comment]
Right, I already have a PR open on our project for such a workaround.  I guess another point of your example is also that this has nothing to do with _listings.sty_ and mathescape; it's much more basic than I had thought, meaning that the issue title as well as the initial comment are a bit misleading.

**Comment by xworld21** [comment]
Before you close the issue, I'd like to add a few use cases and motivations for an (optional) plain text `<mtext>`:
- degrading to plain text works well in some situations, but cannot be done from LaTeX; for instance, using `\ref` inside an equation degrades well to text (you get the reference number) but currently produces an `<a>` tag;
- MathJax is sometimes needed for its accessibility features, even when the browser supports MathML natively;
- many EPUB readers use MathJax to render MathML (see the [MathJax docs](https://docs.mathjax.org/en/v2.7-latest/misc/epub.html#mathml-support-on-epub3-reading-systems)); to my memory, that includes some physical e-readers, but I can't find a definite source at the moment.

What about running the `mtext` content through an XSLT filter (say a `text` mode) to remove all the tags, with warnings when the content is likely being destroyed (e.g. more MathML, images without alt text), and having all of that behind an XSLT parameter? A sort of optional MathJax compatibility mode.

**Comment by dginev** [comment]
@xworld21 I don't follow why it "cannot be done from LaTeX".

For example, we just added support for the math mode hooks (see #1431 ), so you could solve this in TeX by binding `\emph` to `\relax` locally when math mode starts, to avoid the mathjax-hostile mtext.

Here's a minimal document that works today:
```tex
\documentclass{article}
\usepackage{amsmath}
\everymath{\let\emph\relax}
\begin{document}
\emph{emph1} $x=\text{\emph{noemph}}$ \emph{emph2}.
\end{document}
```

This works well with both pdflatex and latexml from my tests. And if you only want it available for the XML side, `\everymath` can be loaded only when latexml is running via `\iflatexml`:
```tex
\documentclass{article}
\usepackage{amsmath}
\usepackage{latexml}
\iflatexml
  \everymath{\let\emph\relax}
\fi
\begin{document}
\emph{emph1} $x=\text{\emph{noemph}}$ \emph{emph2}.
\end{document}
```

which will keep it emphasized in the PDF, but remove the markup in the generated XML/HTML.

I'll restate my preference that we should not go out of our way to break compatibility with MathJax, but this is a case that calls for a workaround, rather than a change of default, until they conform better to the MathML spec.

**Comment by xworld21** [comment]
You can't do that with `\ref`, though, you won't get any text at all (...or can you? I admit I haven't thought about it too hard).

**Comment by dginev** [comment]
Indeed, `\ref` is harder to bind since it has to execute custom Perl code. For those cases the workaround is to define a new perl macro `\plainref` and bind `\let\ref\plainref` in everymath, which is arguably more work. But doable if you really cared about that arguably rather rare case of using a `$\text{... \ref{` setup. It's a lot easier to claim Bruce's workaround is the more correct one for references - just end the formula and put it in the text.

**Comment by henrikt-ma** [comment]
A problem with the `\let\emph\relax` is of course also that it removes the italic style.  Going back to my original comment of not producing the `<em>` tag as a workaround for MathJax, I did some additional testing now that further illustrates the difficulty of the whole situation:
- Firefox w/o MathJax: Best result
- Chrome + MathJax: Looks OK, text still in italics
- Safari w/o MathJax: Not good, not respecting `mathvariant="italic"`

So, fixing MathJax breaks Safari, possibly due to exposing an issue with Safari, but still…  It makes it easy for me to agree that not rocking the boat too much really appears as the most sane approach.

**Comment by xworld21** [comment]
@henrikt-ma Thanks for checking! So the "solution" is worse than the problem.

I suspect I'll end up implementing a text mode filter for my colleagues (the `\ref` inside `align` and `cases` has appeared a few times in our materials, and people had to reshuffle all the text around the problem), if we get sensible results I might come back with a new issue and some real life experience.

To be clear, I wasn't suggesting to change the default, only to add an XSLT parameter to switch all `<mtext>` content to plain text, with the `\ref` use case in mind and no font variants involved – a different kind of situation.

**Comment by dginev** [comment]
Closing here for now, since we have no clear course of action, ideally folks here also ping the MathJax folks on the topic.

**Comment by teepeemm** [comment]
@dginev  You mentioned a `\plainref` above.  How would that work?  I could also use `\ref*` with hyperref instead of `\plainref` (I think), but it seems that we don't respect the `*` option.  So maybe this would be better phrased as: "would it be possible (and not too difficult) to implement the `*` option for `\ref`?"

**Comment by dginev** [comment]
Respecting the `*` of `\ref` is a reasonable enhancement request for latexml, certainly. That's probably the best way to go really. Feel free to open a new issue!
